PubSub Spring Tutorial

anguillanneuf opened the pull request:
This tutorial is mostly done. 😃 I would like to ask a few folks (@elefeint, @kir-titievsky, @ChengyuanZhao, @hongalex, @feywind, @davidcavazos, @soyrice) to review in addition to the review I will get from the community tutorial team. Thanks!

The writeup of the tutorial is in [`index.md`](https://github.com/GoogleCloudPlatform/community/blob/pubsub-spring/tutorials/pubsub-spring-dedup-messages/index.md).

---
+ [Community Tutorial Style Guide](https://cloud.google.com/community/tutorials/styleguide).

hongalex commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,244 @@
+---
+title: Deduplicate Pub/Sub Messages Using Dataflow in A Spring Boot Application
```
nit: lowercase A -> "in a Spring Boot Application"

hongalex commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,244 @@
+
+[Binding and Binding Names]: https://github.com/spring-cloud/spring-cloud-stream/blob/master/docs/src/main/asciidoc/spring-cloud-stream.adoc#binding-and-binding-names
+[Common Binding Properties]: https://github.com/spring-cloud/spring-cloud-stream/blob/master/docs/src/main/asciidoc/spring-cloud-stream.adoc#common-binding-properties
+ [Google Cloud Platform services that integrate with Spring]: https://spring.io/projects/spring-cloud-gcp
```
nit: extra whitespace in here and L243

hongalex commented on tutorials/pubsub-spring-dedup-messages/pubsub-spring/src/main/resources/static/index.html:
```diff
@@ -0,0 +1,26 @@
+
+  <div class="field">
+    <span class="label">Key:</span>
+    <input  name="key" />
```
nit: extra whitespace

hongalex commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,244 @@
+   * From the **Service account** list, select **New service account**.
+   * In the **Service account name** field, enter a name.
+   * From the **Role** list, select **Project > Owner**.
+   * Click **Create**. A JSON file that contains your key downloads to your computer.
```
I think the phrasing is a bit strange. Prefer something like "Save this JSON file to a location on your computer".

hongalex commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,244 @@
+   > *Note:* The **Role** field authorizes your service account to access resources.
+   > You can view and change this field later by using the
+   > [GCP Console IAM page].
+   > If you are developing a production app, specify more granular permissions than **Project > Owner**.
```
Would it be difficult to specify what roles this SA needs in the tutorial? We generally dissuade people from giving the Owner role.

anguillanneuf replied:
I'd leave it as Project Owner.

- There is a link in the notes beneath that tells users how to achieve granular control. 
- Say they want to check out Monitoring, or modify their Dataflow pipeline to write to GCS. Having Project Owner just seems easier.

hongalex replied:
I'm still of the opinion that if we can be more specific, the better. My reasoning is not just about best practices, but also making sure we teach people about the appropriate permissions needed to accomplish what's needed in this tutorial (and eventually in production).

anguillanneuf replied:
I think while it's good to teach people what's needed in production, it becomes a distraction for the tutorial. 
For examples, choosing between `roles/storage.objectAdmin` and `roles/storage.admin`, or `roles/pubsub.admin` versus `roles/pubsub.publisher` could raise more questions, which are best answered by the individual service's IAM pages like [this](https://cloud.google.com/storage/docs/access-control/iam-roles) and [this](https://cloud.google.com/pubsub/docs/access-control). 
I don't see other tutorials do so either, including the ones on cloud site.

anguillanneuf replied:
Alex found me some examples that do grant granular roles. I updated the instructions.

hongalex commented on tutorials/pubsub-spring-dedup-messages/index.md:
````diff
@@ -0,0 +1,244 @@
+mvn spring-boot:run
+```
+
+Observe that your app has started successfully by pointing your browser to `localhost:8080`. You should be able to send messages using the form there. At this point, Spring will have automatically created the Cloud Pub/Sub topic that you have specified in `application.properties` to publish your message to. You can view Publish Message Request Count and Publish Message Operation Count in the topi ...
````
nit: s/Observe/Check

hongalex commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,244 @@
+
+## Start a Cloud Dataflow Job to Deduplicate Pub/Sub Messages
+
+As a middle process that takes data from a Cloud Pub/Sub topic and publishes processed data to another Cloud Pub/Sub topic, Cloud Dataflow achieves exactly once stream processing by asking the input stream for an `idAttribute`. Using the user-defined key name as this unique record identifier name, Cloud Dataflow can avoid processing messages of the same key multiple times.
```
This is a little confusing. Is `idAttribute` the unique record identifier name? If so, what is the "user-defined key name"?

ToddKopriva commented:
I'll do my editorial review and production pass when Alex's review is complete and his suggestions have been resolved.

soyrice commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,263 @@
+
+
+## Objectives
+- Configure a Spring Boot application to use [Cloud Pub/Sub] as a message broker.
```
Are these brackets for Markdown links?

anguillanneuf replied:
Yes.

soyrice replied:
Will the links populate automatically?

anguillanneuf replied:
Oh yes, all the links are in the bottom of the file. It's to avoid typing the same link for the same phrase over and over again.

soyrice commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,263 @@
+   For more information, see
+   [Creating and managing service accounts].
+
+1. Set your `GOOGLE_APPLICATION_CREDENTIALS` environment variable to point to your service account key file.
```
The following command directly sets the variable to the file, so we could probably shorten this to "Set the `GOOGLE_APPLICATION_CREDENTIALS` to your service account key file."

soyrice commented on tutorials/pubsub-spring-dedup-messages/index.md:
````diff
@@ -0,0 +1,263 @@
+   ```shell script
+   export GOOGLE_APPLICATION_CREDENTIALS=path/to/your/credentials.json
+   ```
+   If you have not done it, set your `PROJECT_NAME` environment variable to your GCP project.
````
This step is already in the previous section with the gcloud instructions, so I don't think we need to repeat it here.

anguillanneuf replied:
If they use the UI to complete the previous step, they would have missed the setup for `PROJECT_NAME`.

soyrice replied:
Gotcha - SG

soyrice commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,263 @@
+
+## Bind Cloud Pub/Sub to Your Spring Boot Application
+
+Spring makes use of Spring Cloud Stream binders to send data to a Cloud Pub/Sub topic and receive data from a Cloud Pub/Sub subscription.
```
We should refer to "Cloud Pub/Sub" as just "Pub/Sub" now

soyrice commented on tutorials/pubsub-spring-dedup-messages/index.md:
````diff
@@ -0,0 +1,263 @@
+```
+
+### Specify a Cloud Pub/Sub topic for the source
+Spring can find the Cloud Pub/Sub topic that your code should publish data to when you provide a topic name for your source in `application.properties` using an output binder. Here, the output binder's binding name is `sendMessagesForDeduplication-out-0`. For more information, see [Binding and Binding Names].
````
This sentence has a lot of parts so it's kind of hard to follow. I'd recommend splitting it up into at least two sentences. It may be helpful to start this sentence with the section's objective: "To specify the source Pub/Sub topic..."

soyrice commented on tutorials/pubsub-spring-dedup-messages/index.md:
````diff
@@ -0,0 +1,263 @@
+spring.cloud.stream.bindings.sendMessagesForDeduplication-out-0.destination=topicToDataflow
+```
+
+Similarly, your application will know how to receive data from a Cloud Pub/Sub subscription if you specify 1). a Spring Cloud Stream sink and 2). a pair of Cloud Pub/Sub topic and subscription.
````
I'd recommend using the list formatting here. You can emphasize that both list items are required with a phrase like: "...if you specify both of the following:"

soyrice commented on tutorials/pubsub-spring-dedup-messages/index.md:
````diff
@@ -0,0 +1,263 @@
+```
+
+### Specify a Cloud Pub/Sub topic and subscription for the sink
+Spring can find the Cloud Pub/Sub subscription that your code should receive data from when you provide a subscription name for your sink in `application.properties` using an input binder. Here, the input binder's binding name is `receiveDedupedMessagesFromDataflow-in-0`. Because a subscription cannot exist without a topic, a topic must also be specified. The input binder's destination is your Pu ...
````
Similar comment as before. It may be useful to start this sentence with the section's objective.

Is the `receiveDedupedMessagesFromDataflow-in-0` binder relevant only if you're following the tutorial or is it a keyword that applies to all applications can use to get the subscription name? This also applies to other things in code font, like `receiveDedupedMessagesFromDataflow` and `topicFromDataflow`.

soyrice commented on tutorials/pubsub-spring-dedup-messages/index.md:
````diff
@@ -0,0 +1,263 @@
+mvn spring-boot:run
+```
+
+Check that your app has started successfully by pointing your browser to `localhost:8080`. You should be able to send messages using the form there. At this point, Spring will have automatically created the Cloud Pub/Sub topic that you have specified in `application.properties` to publish your message to. You can view Publish Message Request Count and Publish Message Operation Count in the topic  ...
````
This paragraph is hard to follow and may be easier to follow as a list of steps the reader can do to verify that the app is working.

soyrice commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,263 @@
+
+## Start a Cloud Dataflow Job to Deduplicate Pub/Sub Messages
+
+As a middle process that takes data from a Cloud Pub/Sub topic and publishes processed data to another Cloud Pub/Sub topic, Cloud Dataflow achieves exactly once stream processing by asking the input stream for an `idAttribute`. The `idAttribute` can be a Pub/Sub message attribute. Cloud Dataflow will avoid processing messages that have the same values in this `idAttribute` repeatedly.
```
It's not clear to me how being a middle process lets Dataflow achieves exactly-once processing. This paragraph suggests that it achieves exactly-once processing by getting the `idAttribute` from the input stream. 

We should also use "Dataflow" instead of "Cloud Dataflow" now.

soyrice commented on tutorials/pubsub-spring-dedup-messages/index.md:
````diff
@@ -0,0 +1,263 @@
+     --idAttribute=key \
+     --runner=DataflowRunner"
+```
+You can check the job's progress in the [Cloud Console for Dataflow]. Wait a few minutes for the job status to become "running". Then issue `Ctrl+C` to stop the program locally. Stopping the program locally does not affect the Dataflow job running in the Cloud.
````
How come "running" is quotes?

anguillanneuf replied:
I was quoting this screenshot.

![Screen Shot 2020-02-13 at 5 38 24 PM](https://user-images.githubusercontent.com/9518298/74577880-6a6c2a00-4f46-11ea-8561-ffece3aad490.png)

I think I will add this screenshot. Wdyt?

soyrice replied:
The screenshot could help. In either case, I think **Status** and **Running** should be in title case + bold since they refer to elements in the UI. You may need to rephrase the sentence a bit.

anguillanneuf replied:
Sounds good! Will add a screenshot and bold + upper case text here.

davidcavazos commented on tutorials/pubsub-spring-dedup-messages/pubsubio-dedup/pom.xml:
```diff
@@ -0,0 +1,180 @@
+  <version>1.0</version>
+
+  <properties>
+    <maven.compiler.source>1.8</maven.compiler.source>
```
Can we also use version `11`?

davidcavazos commented on tutorials/pubsub-spring-dedup-messages/pubsubio-dedup/pom.xml:
```diff
@@ -0,0 +1,180 @@
+    <maven.compiler.target>1.8</maven.compiler.target>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+
+    <beam.version>2.18.0</beam.version>
```
I think `2.19.0` is out already?

```
# Check your plugin versions.
mvn versions:display-plugin-updates

# Check your dependency versions.
mvn versions:display-dependency-updates
```

davidcavazos commented on tutorials/pubsub-spring-dedup-messages/pubsub-spring/pom.xml:
```diff
@@ -0,0 +1,37 @@
+    <dependency>
+      <groupId>org.springframework.cloud</groupId>
+      <artifactId>spring-cloud-gcp-pubsub-stream-binder</artifactId>
+      <version>1.2.0.RELEASE</version>
```
I've heard it's good practice to keep [versions as variables](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/88ae90245aa8a6f6ddc84f8fdc47ada35326aeb3/dataflow/encryption-keys/pom.xml#L33-L37) in the `<properties>` section.

anguillanneuf replied:
This `pom.xml` is really short. I would like to keep it simple like it is. :)

davidcavazos commented on tutorials/pubsub-spring-dedup-messages/pubsub-spring/pom.xml:
```diff
@@ -0,0 +1,37 @@
+    </dependency>
+  </dependencies>
+
+</project>
```
Nit: newline at EOF

davidcavazos commented on tutorials/pubsub-spring-dedup-messages/pubsub-spring/src/main/java/com/google/example/DataEntryController.java:
```diff
@@ -0,0 +1,53 @@
+  @PostMapping("/sendData")
+  public RedirectView mainPage(
+    @RequestParam("data") String data, @RequestParam("key") String key) {
+    System.out.println("Sending data: " + data + " of key " + key + "..");
```
You might want to consider `String.format()` or `System.out.printf()`?

```
System.out.println(String.format("Sending data: {} of key {}", data, key));
System.out.printf("Sending data: %s of key %s", data, key);
```

They're all a bit inconsistent :/

anguillanneuf replied:
What's the benefit of switching to `String.format`?

davidcavazos commented on tutorials/pubsub-spring-dedup-messages/pubsub-spring/src/main/resources/static/index.html:
```diff
@@ -0,0 +1,26 @@
+
+
+</body>
+</html>
```
Nit: newline at EOF

davidcavazos commented on tutorials/pubsub-spring-dedup-messages/pubsubio-dedup/pom.xml:
```diff
@@ -0,0 +1,180 @@
+      <version>${beam.version}</version>
+    </dependency>
+
+    <dependency>
```
Are you using `beam-sdks-java-maven-archetypes-examples` for anything like the `WordCount` or mobile gaming? If not consider removing this.

anguillanneuf commented on tutorials/pubsub-spring-dedup-messages/index.md:
````diff
@@ -0,0 +1,260 @@
+   ```shell script
+    export PROJECT_NAME=$(gcloud config get-value project)
+   ```
+
````
Spring automatically creates them if they do not exist. I wrote that down in the tutorial this time. :)

anguillanneuf commented on tutorials/pubsub-spring-dedup-messages/index.md:
````diff
@@ -0,0 +1,260 @@
+1. In the [Cloud Console for Dataflow], select the Dataflow job and stop it. Cancel the pipeline instead of draining it. Dataflow will automatically delete the subscription to the topic `topicFromDataflow` during this process. 
+1. Delete the subscription followed by the topics in the [Cloud Console for Pub/Sub] or via the command line.
+```shell script
+gcloud pubsub subscriptions delete topicFromDataflow.subscriptionFromDataflow
````
You shouldn't need `--topic` to delete a subscription. I just tried :)
Spring creates a subscription that's a concatenation of the user-provided topic name and subscription name, in this case, it is `topicFromDataflow.subscriptionFromDataflow`.

anguillanneuf commented on tutorials/pubsub-spring-dedup-messages/index.md:
````diff
@@ -0,0 +1,260 @@
+```
+
+### Specify a Cloud Pub/Sub topic for the source
+Spring can find the Cloud Pub/Sub topic that your code should publish data to when you provide a topic name for your source in `application.properties` using an output binder. Here, the output binder's binding name is `sendMessagesForDeduplication-out-0`. For more information, see [Binding and Binding Names].
````
I'm not sure about this one. Pinging @ToddKopriva to see if this becomes a problem when published.

anguillanneuf commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,260 @@
+
+
+## Objectives
+- Configure a Spring Boot application to use [Cloud Pub/Sub] as a message broker.
```
I added an intro paragraph :)

anguillanneuf commented:
@hongalex I updated the note section and added my [comment](https://github.com/GoogleCloudPlatform/community/pull/1152#discussion_r380911480) on this.

davidcavazos commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -212,22 +217,22 @@ In [DedupPubSub.java](pubsubio-dedup/src/main/java/com/google/example/DedupPubSu
 
 ## Observe the Results
-Publish a few messages of different keys via the web host and observe deduplicated messages in your terminal.
+Publish a few more messages of different keys (i.e. `idAttributes`) via the web host and observe messages of the same key arrive only once in your terminal.
```
Can you add examples of key-message pairs and the expected output? Maybe something like:

```
key: "111", message: "hi"     # processed since key 111 has not been seen before
key: "222", message: "hello"  # processed
key: "111", message: "ignore" # ignored since key 111 has been seen before
```

anguillanneuf replied:
I added a paragraph beneath a screenshot of the output to describe the expected output.

davidcavazos commented on tutorials/pubsub-spring-dedup-messages/pubsub-spring/pom.xml:
```diff
@@ -15,12 +15,11 @@
   <properties>
-    <maven.compiler.source>1.8</maven.compiler.source>
-    <maven.compiler.target>1.8</maven.compiler.target>
+    <maven.compiler.source>1.11</maven.compiler.source>
```
I think you can specify `11` only instead of `1.11`.

anguillanneuf replied:
I'm using Java 1.8 instead of Java 11.

davidcavazos commented on tutorials/pubsub-spring-dedup-messages/pubsubio-dedup/pom.xml:
```diff
@@ -24,11 +24,11 @@
   <properties>
-    <maven.compiler.source>1.8</maven.compiler.source>
-    <maven.compiler.target>1.8</maven.compiler.target>
+    <maven.compiler.source>1.11</maven.compiler.source>
```
`1.11` -> `11`

anguillanneuf replied:
I'm using Java 1.8 instead of Java 11.

elefeint commented on tutorials/pubsub-spring-dedup-messages/pubsubio-dedup/pom.xml:
```diff
@@ -0,0 +1,174 @@
+
+  <repositories>
+    <repository>
+      <id>apache.snapshots</id>
```
Do you need this repo here? This project relies on a released version of Beam, not on a snapshot.

anguillanneuf commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,284 @@
+   * From the **Service account** list, select **New service account**.
+   * In the **Service account name** field, enter a name.
+   * From the **Role** list, select **PubSub > Admin** and **Dataflow > Admin**.
+   * Click **Create**. Save this JSON file to a location on your computer.
```
You might be looking at a different page than the one appointed in the tutorial. 

![Screen Shot 2020-02-19 at 10 47 00 AM](https://user-images.githubusercontent.com/9518298/74872566-4e83d200-5312-11ea-9098-9d701815f68d.png)

anguillanneuf commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,284 @@
+[Common Binding Properties]: https://github.com/spring-cloud/spring-cloud-stream/blob/master/docs/src/main/asciidoc/spring-cloud-stream.adoc#common-binding-properties
+[Google Cloud Platform services that integrate with Spring]: https://spring.io/projects/spring-cloud-gcp/
+[Dataflow templates]: https://cloud.google.com/dataflow/docs/guides/templates/overview/
+[binders]: https://github.com/spring-cloud/spring-cloud-stream/blob/master/docs/src/main/asciidoc/spring-cloud-stream.adoc#the-binder-abstraction
```
Removed the URL.
Added dependency code in the body of the tutorial.

anguillanneuf commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,284 @@
+
+   * From the **Service account** list, select **New service account**.
+   * In the **Service account name** field, enter a name.
+   * From the **Role** list, select **PubSub > Admin** and **Dataflow > Admin**.
```
I tried a service account that have only PubSub Admin and Dataflow Admin. It worked actually.

anguillanneuf commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,284 @@
+   or via the `gcloud` command line tool.
+   Here is how to do it through the *Create service account key* page.
+
+   * From the **Service account** list, select **New service account**.
```
It is under "IAM and Admin", but the link provided above actually takes the user straight here:

![Screen Shot 2020-02-19 at 10 47 00 AM](https://user-images.githubusercontent.com/9518298/74872748-b6d2b380-5312-11ea-97b8-86b056f92a34.png)

anguillanneuf commented on tutorials/pubsub-spring-dedup-messages/index.md:
```diff
@@ -0,0 +1,284 @@
+
+Similarly, your application can receive data from a Pub/Sub subscription if you specify both of the following: 
+1. a Spring Cloud Stream sink
+1. a pair of Pub/Sub topic and subscription
```
I rephrased the paragraph below to say "a destination name and an optional group name". 

I also rephrased the sentence here to avoid misleading information.

anguillanneuf commented:
Friendly ping to @ToddKopriva for review 😃

ToddKopriva commented:
I've been waiting for the dust to settle from everyone else's reviews. Is it itme for the editorial review?

anguillanneuf commented:
@ToddKopriva Yes, thanks for waiting! :)

ToddKopriva commented:
Thanks for this excellent new tutorial. I'll publish this to cloud.google.com with a set of other documents later this week.
